lib/Config/Config.php

```diff
@@ -0,0 +1,333 @@
+		'ldapPagingSize'                 => 500,
+		'ldapExperiencedAdmin'           => 0,
+		'ldapDynamicGroupMemberURL'      => '',
+		'ldapIgnoreNamingRules'          => false
+	];
+
+	public function __construct(array $data) {
+		$this->parseData($data);
+	}
+
+	/**
+	 * @return string the configuration prefix
+	 */
+	public function getId() {
+		return $this->id;
+	}
+
+	public function setId($id) {
+		$this->id = $id;
+	}
+
+	public function getData() {
+		return $this->data;
+	}
+
+	/**
+	 * @param string $name
+	 *
+	 * @return mixed|null
+	 */
+	public function __get($name) {
+		if (isset($this->data[$name])) {
+			return $this->data[$name];
+		}
+		return null;
+	}
+
+	/**
+	 * @param string $name
+	 *
+	 * @param mixed $value
+	 */
+	public function __set($name, $value) {
+		$this->setConfiguration([$name => $value]);
+	}
+
+	public function jsonSerialize() {
+		$data = [];
+		foreach ($this->data as $key=>$value) {
+			switch ($key) {
+				case 'ldapAgentPassword':
+					$value = \base64_encode($value);
+					break;
+				//following options are not stored but detected, skip them
+				case 'ldapIgnoreNamingRules':
+				case 'hasPagedResultSupport':
+				case 'ldapUuidUserAttribute':
+				case 'ldapUuidGroupAttribute':
+					continue 2;
+			}
+			if ($value === null) {
+				$value = '';
+			}
+			$data[$key] = $value;
+		}
+		return $data;
+	}
+
+	/**
+	 * set LDAP configuration with values delivered by an array, not read
+	 * from configuration. It does not save the configuration! To do so, you
+	 * must call saveConfiguration afterwards.
+	 * @param array $config array that holds the config parameters in an associated
+	 * array
+	 * @param array &$applied optional; array where the set fields will be given to
+	 * @return false|null
+	 */
+	public function setConfiguration($config, &$applied = null) {
+		if (!\is_array($config)) {
+			return false;
+		}
+
+		foreach ($config as $inputKey => $val) {
+			if (\array_key_exists($inputKey, $this->data)) {
+				$key = $inputKey;
+			} else {
+				continue;
+			}
+
+			switch ($key) {
+				case 'ldapAgentPassword':
+					$this->setRawValue($key, $val);
+					break;
+				case 'ldapBase':
+				case 'ldapBaseUsers':
+				case 'ldapBaseGroups':
+				case 'ldapAttributesForUserSearch':
+				case 'ldapAttributesForGroupSearch':
+				case 'ldapUserFilterObjectclass':
+				case 'ldapUserFilterGroups':
+				case 'ldapGroupFilterObjectclass':
+				case 'ldapGroupFilterGroups':
+				case 'ldapLoginFilterAttributes':
+					$this->setMultiLine($key, $val);
+					break;
+				case 'homeFolderNamingRule':
+					$trimmedVal = \trim($val);
+					if ($trimmedVal !== '' && \strpos($val, 'attr:') === false) {
+						$val = 'attr:' . $trimmedVal;
+					}
+				// no break
+				default:
+					$this->setValue($key, $val);
+			}
+			if (\is_array($applied)) {
+				$applied[] = $inputKey;
+			}
+		}
+		return null;
+	}
+
+	public function getDefaults() {
+		$defaultConfig = new self([]);
```

Comment:
A `Config` object creating another `Config` object to get its data doesn't look good. I think this is another reason to proceed with the `ConfigFactory` idea: you can easily do the same just by `$configFactory->getFreshConfig()->getData();` or even something like `$configFactory->getFreshConfig()->isEqualTo($config);` if we're interested in comparing configurations.

Reply:
@jvillafanez I'd prefer not to refactor too much as rebasing https://github.com/owncloud/user_ldap/pull/210 on top of this PR will give me enough headache already.

Reply:
Ok. Still, I think we should open an issue to refactor and cleanup these things after eveything is merged.